### PR TITLE
[DeepSeek V4.1] an FP8 Tensor Core path for the FP4 grouped indexer

### PR DESCRIPTION
**Pull request by zhangxiaolei123456**
## Motivation

DeepSeek-V4.1 stores indexer keys in a compact FP4 cache representation and quantizes indexer queries to the same representation at the grouped CUDA entry: 128 E2M1 values packed into 64 bytes, plus four UE8M0 scales covering 32 elements each. During target verification, multiple query rows from the same request score the same compressed K sequence.

The existing SM90 Triton fallback expands packed FP4 values and computes the logits independently for every query row. This leaves two optimization opportunities on Hopper:

1. Reuse each decoded K tile across the request's verification rows.
2. Avoid decoding the same Q rows again for every K tile.

Hopper does not provide native MXFP4 MMA. This PR therefore converts packed E2M1 values to E5M2 in shared memory and uses SM90 FP8 WGMMA. The FP4 block scales are factored into a common power-of-two scale and restored after the tensor-core accumulation.

## Summary

- Optimize FP4 scale decoding in the existing SM90 Triton fallback.
- Add an opt-in grouped CUDA indexer for request-major speculative rows.
- Decode one K tile and reuse it across all query rows in the request group.
- Keep up to six decoded Q rows resident in shared memory across 2, 4, or 8 K tiles.
- Preserve the existing BF16 rounding points, ReLU, per-head weighting, and final head reduction semantics.
- Add an SM90 unit test comparing the grouped CUDA path against the existing Triton implementation.

The grouped path remains opt-in in this PR. It is not enabled globally because Triton is still faster for small workloads.

## Computation

For query row `b`, head `h`, and compressed position `j`, the indexer computes:

```text
dot[b,h,j] = BF16(sum_d Q[b,h,d] * K[j,d])
score       = ReLU(dot)
weighted    = BF16(score * weight[b,h])
logit[b,j]  = BF16(sum_h weighted[b,h,j])
```

The production shape is:

```text
heads    = 64
head_dim = 128
K tile   = 64 compressed positions
```

One tile is therefore a `[64, 128] x [128, 64]` matrix multiplication, or 1,048,576 floating-point operations per query row.

## SM90 Mapping

The CUDA kernel uses:

```text
GMMA::MMA_64x64x32_F32E5M2E5M2_SS_TN
```

Four WGMMA operations cover the K dimension of 128. One 128-thread warpgroup owns one query row, and a 512-thread CTA contains four warpgroups.

For every 32-element FP4 scale group, where the UE8M0 byte uses exponent bias
127:

```text
FP4 value = E2M1(code) * 2^(ue8m0_exponent - 127)
          = E5M2(adjusted_code) * 2^(common_exponent - 127)
```

The conversion uses a small E2M1-to-E5M2 lookup table followed by an exact E5M2 exponent-field adjustment. The common Q and K scales are multiplied into the FP32 accumulator before applying the required BF16 rounding.

## Q Residency

For a six-row request group and one 64-position K tile, the conversion work is:

```text
K: 64 * 128     =  8,192 values
Q: 6 * 64 * 128 = 49,152 values
```

Q accounts for 85.7% of the converted values if it is decoded for every K tile. The multi-tile variants decode Q once, retain six E5M2 Q matrices in shared memory, and process multiple K tiles:

```text
TILES_PER_CTA = 2, 4, or 8
```

With eight K tiles per CTA, Q conversion work per tile is reduced by 87.5%.

The H20 dispatch heuristic is:

```text
total L64 tiles < 128   -> 1 tile/CTA
total L64 tiles >= 128  -> 2 tiles/CTA
total L64 tiles >= 256  -> 4 tiles/CTA
total L64 tiles >= 1024 -> 8 tiles/CTA
```

Multi-tile dispatch also requires enough tiles in each request group. Groups larger than six retain the single-tile path.

CTA-wide barriers are used during Q preparation. An earlier named-barrier implementation produced racecheck hazards despite passing output checks and was discarded.

## Resource Usage

Measured from the compiled SM90 kernels:

| Variant | Registers/thread | Dynamic shared memory | Resident CTAs/SM | Spills |
|---|---:|---:|---:|---:|
| Single-tile baseline | 64 | 46,848 bytes | 2 | 0 |
| Multi-tile Q-resident | 60 | 63,744 bytes | 2 | 0 |

## Correctness and Safety Validation

Measured on an NVIDIA H20.

- 23 shape/configuration cases:
  - widths from 1 to 1025;
  - query group sizes 2, 3, 4, 5, 6, 7, and 9;
  - complete and partial final groups;
  - page sizes 16 and 64;
  - compression ratios 1 and 2;
  - random physical cache slots;
  - per-row visible lengths, including zero;
  - padded cache and output strides;
  - output guard regions.
- Every applicable explicit variant and automatic dispatch matched the single-tile CUDA baseline bit for bit.
- An independent integer-dot oracle matched bit for bit.
- The real Python/JIT wrapper was tested with:
  - BF16-to-FP4 Q quantization;
  - partial request groups;
  - CUDA Graph capture and replay;
  - width-zero no-launch behavior.
- `compute-sanitizer` results over all 23 cases:
  - memcheck: zero errors;
  - racecheck: zero hazards;
  - synccheck: zero errors.
- Repository `clang-format` and `git diff --check` passed.

The grouped CUDA unit test also compares against the Triton fallback using production RNE FP4 quantization.

### CUDA vs. Existing Triton Fallback

The following comparison uses nine timing rounds, with 20 calls per CUDA Graph replay. Both paths receive identical FP4-representable BF16 Q, positive random head weights, and the same packed K cache. Random physical slots, padded cache strides, compression ratio two, and varying visible lengths are included.

`CUDA total` includes the additional Q packing required by the grouped CUDA path. Triton consumes BF16 Q directly. Model-level fake quantization is shared upstream and excluded from both measurements.

| Query rows | Compressed length | Triton | CUDA logits only | CUDA total | Triton / CUDA total |
|---:|---:|---:|---:|---:|---:|
| 5 | 65 | 6.80 us | 27.64 us | 31.96 us | 0.21x |
| 5 | 512 | 6.95 us | 27.92 us | 31.87 us | 0.22x |
| 23 | 512 | 12.87 us | 28.63 us | 36.08 us | 0.36x |
| 95 | 512 | 35.41 us | 38.32 us | 58.65 us | 0.60x |
| 383 | 512 | 131.86 us | 90.39 us | 155.14 us | 0.85x |
| 23 | 4096 | 67.95 us | 57.71 us | 65.11 us | 1.04x |
| 95 | 4096 | 251.95 us | 147.86 us | 165.33 us | 1.52x |
| 383 | 4096 | 987.33 us | 480.13 us | 554.84 us | 1.78x |
| 95 | 16384 | 978.87 us | 476.26 us | 493.72 us | 1.98x |

The CUDA path wins for the tested longer sequences and larger batches. Triton wins for all tested length-512 cases. At 383 rows and length 512, the CUDA logits kernel alone is 1.46x faster, but Q packing removes that advantage.

Across these nine synthetic cases:

- CUDA and Triton finite logits matched exactly;
- `-inf` masks matched exactly;
- the new and old CUDA paths matched bit for bit;
- selected top-k sets matched.

The synthetic equality result does not claim numerical identity for every possible arbitrary scale pattern.

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :x: [Run #35328484638](https://github.com/sgl-project/sglang/actions/runs/35328484638)
Latest PR Test (Extra): :x: [Run #35328484453](https://github.com/sgl-project/sglang/actions/runs/35328484453)
Latest PR Test (AMD ROCm 10): :x: [Run #35328484576](https://github.com/sgl-project/sglang/actions/runs/35328484576)

### CODE REVIEW

**Comment by yuan-luo** [comment]
Please fix the lint first.
